Let's `#[expect]` some lints: Stabilize `lint_reasons` (RFC 2383)

xFrednet opened the pull request:
Let's give this another try! The [previous stabilization attempt](https://github.com/rust-lang/rust/pull/99063) was stalled by some unresolved questions. These have been discussed in a [lang team](https://github.com/rust-lang/lang-team/issues/191) meeting. The last open question, regarding the semantics of the `#[expect]` attribute was decided on in https://github.com/rust-lang/rust/issues/115980

I've just updated the [stabilization report](https://github.com/rust-lang/rust/issues/54503#issuecomment-1179563964) with the discussed questions and decisions. Luckily, the decision is inline with the current implementation.

This hopefully covers everything. Let's hope that the CI will be green like the spring.

fixes #115980
fixes #54503

--- 

For the release note author:

There are two Clippy lints which could be noteworthy:
* [`clippy::allow_attributes`](https://rust-lang.github.io/rust-clippy/master/index.html#/allow_attributes) to restrict allow attributes in favor of `#[expect]` or to migrate `#[allow]` attributes to `#[expect]`
* [`clippy::allow_attributes_without_reason`](https://rust-lang.github.io/rust-clippy/master/index.html#/allow_attributes_without_reason) To require a reason for `#[allow]` attributes

Both are *allow-by-default* lints in the `restriction` category.

---

r? @wesleywiser

Tacking Issue: https://github.com/rust-lang/rust/issues/54503
Stabilization Report: https://github.com/rust-lang/rust/issues/54503#issuecomment-1179563964
Documentation Update: https://github.com/rust-lang/reference/pull/1237

<!--
For Clippy:

changelog: [`allow_attributes`]: Is now available on stable, since the `lint_reasons` feature was stabilized
changelog: [`allow_attributes_without_reason`]: Is now available on stable, since the `lint_reasons` feature was stabilized
-->

---

Roses are red,
Violets are blue,
Let's expect lints,
With reason clues

xFrednet committed:
RFC 2383: Stabilize `lint_reasons` :tada:

xFrednet committed:
RFC 2383: Stabilize `lint_reasons` in Clippy :paperclips:

xFrednet committed:
RFC 2383: Stabilize `lint_reasons` in Miri

xFrednet committed:
RFC 2383: Update documentation

xFrednet commented:
Alright, looks like most of the CI already agrees with my changes, let's kick off the reviewing process.

r? @wesleywiser

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7861117964/job/21449101094) [(plain)](https://github.com/rust-lang/rust/commit/1f83492c67ea91eae32cc5f622b43601050ea761/checks/21449101094/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=xFrednet
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_eeda8801-b663-4415-9b9d-be83b3f42864
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=a0f99b3dbbf17c265e32aec132984bb0b53575f4
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_eeda8801-b663-4415-9b9d-be83b3f42864
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_eeda8801-b663-4415-9b9d-be83b3f42864
GITHUB_TRIGGERING_ACTOR=xFrednet
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/120924/merge
GITHUB_WORKFLOW_SHA=a0f99b3dbbf17c265e32aec132984bb0b53575f4
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
#12 writing image sha256:d449b693d747d405bbcbff6123aff5df7e8f1b129411b87f1444cc49d40d1e05 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Sun Feb 11 09:46:25 UTC 2024
  network time: Sun, 11 Feb 2024 09:46:25 GMT
  network time: Sun, 11 Feb 2024 09:46:25 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished release [optimized] target(s) in 1.39s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --no-default-features --mini-tests --std-tests`
Using system GCC
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
1 warning: this lint expectation is unfulfilled
-   --> $DIR/expect-tool-lint-rfc-2383.rs:16:11
+   --> $DIR/expect-tool-lint-rfc-2383.rs:15:11
3    |
4 LL | #![expect(rustdoc::missing_crate_level_docs)]

7    = note: `#[warn(unfulfilled_lint_expectations)]` on by default
8 
9 warning: this lint expectation is unfulfilled
9 warning: this lint expectation is unfulfilled
-   --> $DIR/expect-tool-lint-rfc-2383.rs:70:14
+   --> $DIR/expect-tool-lint-rfc-2383.rs:69:14
11    |
12 LL |     #[expect(rustdoc::broken_intra_doc_links)]

14 
15 warning: this lint expectation is unfulfilled
-   --> $DIR/expect-tool-lint-rfc-2383.rs:75:14
-   --> $DIR/expect-tool-lint-rfc-2383.rs:75:14
+   --> $DIR/expect-tool-lint-rfc-2383.rs:74:14
17    |
18 LL |     #[expect(rustdoc::invalid_html_tags)]

20 
21 warning: this lint expectation is unfulfilled
Build completed unsuccessfully in 0:51:08
Build completed unsuccessfully in 0:51:08
-   --> $DIR/expect-tool-lint-rfc-2383.rs:80:14
+   --> $DIR/expect-tool-lint-rfc-2383.rs:79:14
23    |
24 LL |     #[expect(rustdoc::bare_urls)]


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/lints/expect-tool-lint-rfc-2383/expect-tool-lint-rfc-2383.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/lints/expect-tool-lint-rfc-2383/expect-tool-lint-rfc-2383.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lints/expect-tool-lint-rfc-2383.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "/checkout/tests/rustdoc-ui/lints/expect-tool-lint-rfc-2383.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/lints/expect-tool-lint-rfc-2383" "-A" "internal_features" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/lints/expect-tool-lint-rfc-2383/auxiliary"
--- stderr -------------------------------
warning: this lint expectation is unfulfilled
##[warning]  --> /checkout/tests/rustdoc-ui/lints/expect-tool-lint-rfc-2383.rs:15:11
   |
   |
LL | #![expect(rustdoc::missing_crate_level_docs)]
   |
   = note: `#[warn(unfulfilled_lint_expectations)]` on by default

warning: this lint expectation is unfulfilled
warning: this lint expectation is unfulfilled
##[warning]  --> /checkout/tests/rustdoc-ui/lints/expect-tool-lint-rfc-2383.rs:69:14
   |
LL |     #[expect(rustdoc::broken_intra_doc_links)]

warning: this lint expectation is unfulfilled
##[warning]  --> /checkout/tests/rustdoc-ui/lints/expect-tool-lint-rfc-2383.rs:74:14
   |
   |
LL |     #[expect(rustdoc::invalid_html_tags)]

warning: this lint expectation is unfulfilled
##[warning]  --> /checkout/tests/rustdoc-ui/lints/expect-tool-lint-rfc-2383.rs:79:14
   |
   |
LL |     #[expect(rustdoc::bare_urls)]

warning: 4 warnings emitted
------------------------------------------

```

</details>

oli-obk commented on compiler/rustc_lint_defs/src/builtin.rs:
```diff
@@ -616,24 +616,14 @@ declare_lint! {
-    /// It was expected that the marked code would emit a lint. This expectation
-    /// has not been fulfilled.
+    /// The `#[expect]` attribute can be used to create a lint expectation. The
+    /// expectation is fulfilled, if a `#[warn]` attribute at the same location
+    /// would result in a lint emission. If the expectation is unfulfilled,
+    /// because no lint was emitted, this lint will be emitted on the attribute.
```
I think this needs to be rephrased. Isn't the expectation only fulfilled if the lint isn't allowed? Like

```rust
#![allow(some_lint)]

#[expect(some_lint)]
fn foo() {
    something_that_would_emit_some_lint();
}
```

will actually warn, because the `expect` is not fulfilled. The sentence as written signals that the expectation would be fulfilled and this would compile warning free

xFrednet replied:
This question was one of the central questions, discussed in https://github.com/rust-lang/rust/issues/115980. The decision was that expect should act like a lint level. A strong motivator for this was [use case 4](https://hackmd.io/@xFrednet/expect-attr-use-cases#Use-case-4-Suppress-lints-from-CI). The short summary is, that if `#[expect]` is based on the outer lint level, it can be annoying to use with command line arguments, that might only enable some lints in the CI.

I think this model is also easier to explain. If you slap a `#[warn]` attribute there, you would expect a waning, done. You don't need to consider the outer scope.

oli-obk replied:
gotcha, so the impl already reflects that, just the docs were outdated?

xFrednet replied:
Correct. The previous docs actually wanted to express the same thing, but at that time the model wasn't as clear to me, which is why the docs were a bit vague.

xFrednet committed:
RFC 2373: Update tests for rustdoc for `lint_reasons`

oli-obk commented on tests/ui/impl-trait/in-trait/success.rs:
```diff
@@ -1,6 +1,6 @@
 // check-pass
 
-#![feature(lint_reasons)]
+#![cfg_attr(bootstrap, feature(lint_reasons))]
```
you can just remove these attributes from tests. We don't run ui tests on stage 0

oli-obk reviewed (approved):
modulo the bootstrap cfgs in ui tests, the compiler side lgtm

xFrednet commented:
Okay, so anything in the `test/` folder should just have the `#[feature]` removed instead of the `bootstrap` cfgs?

Centri3 commented on src/tools/clippy/clippy_lints/src/attrs.rs:
````diff
@@ -342,6 +340,42 @@ declare_clippy_lint! {
+    /// ```
+    /// Use instead:
+    /// ```rust,ignore
+    /// #![cfg_attr(bootstrap, feature(lint_reasons))]
````
Shouldn't these be commented out? Or removed if what oli said about ui tests applies to doctests, idrk

xFrednet replied:
Maybe? I believe, I tried that in the previous stabilization PR and got an error at that time. The docs are built for the beta and stable channel, to I would guess that they are needed. I'm trying to run the rest right now (Rustc takes so long to compile :sob: Even on the dev machines from rust-lang)

oli-obk replied:
It should be fine to remove them. Clippy is a stage1 tool

wesleywiser commented:
Implementation looks good to me! If I understand the current state correctly, T-lang has signed off on the previously unresolved questions but has not yet FCP'd stabilization. Is that correct? 

I think the next step is then for us to start a stabilization FCP.

xFrednet commented:
> Implementation looks good to me! If I understand the current state correctly, T-lang has signed off on the previously unresolved questions but has not yet FCP'd stabilization. Is that correct?

Now that you say it, I believe this is correct. The [previous stabilization report](https://github.com/rust-lang/rust/issues/54503#issuecomment-1179563964) has been updated and should be sufficient. How do we kick off the FCP, do we ping T-lang in the tacking issue or here? https://github.com/rust-lang/rust/issues/54503

And should I repost the updated stabilization report, just to have it at the bottom?

wesleywiser commented:
Let's go ahead and do the FCP here, just so it's in the same PR that will eventually be merged 🙂 

@rust-lang/lang, now that the implementation questions regarding `#[expect]` have been resolved, we feel the feature is ready for stabilization. @xFrednet has updated the [stabilization report](https://github.com/rust-lang/rust/issues/54503#issuecomment-1179563964) to reflect the current state.

bors commented:
:umbrella: The latest upstream changes (presumably #120881) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tmandry commented:
> Let's go ahead and do the FCP here, just so it's in the same PR that will eventually be merged 🙂
> 
> @rust-lang/lang, now that the implementation questions regarding `#[expect]` have been resolved, we feel the feature is ready for stabilization. @xFrednet has updated the [stabilization report](https://github.com/rust-lang/rust/issues/54503#issuecomment-1179563964) to reflect the current state.

Let's do it!

@rfcbot merge

nikomatsakis commented:
@rfcbot reviewed

I am so ready for this!

Abestanis commented:
Hello, just a gentle bump so this does not get forgotten, would really like to see this land in Rust.

xFrednet commented:
Nice :tada:

My next week will be busy, but I'll rebase this before the end of the month :D

nikomatsakis commented:
So exciting!

On Thu, May 9, 2024, at 6:19 PM, Fridtjof Stoldt wrote:
> 
> 
> Nice 🎉
> 
> My next week will be busy, but I'll rebase this before the end of the month :D
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/rust-lang/rust/pull/120924#issuecomment-2102987607>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABF4ZSRQYJM2WHEVO4LFRDZBOOZVAVCNFSM6AAAAABDDM57FKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMBSHE4DONRQG4>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

xFrednet committed:
`sudo CI=green` && Review changes <3

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9364256518/job/25776884076) [(plain)](https://github.com/rust-lang/rust/commit/c495dbd9c25b2b2f2f31df63da42922feb7cf60a/checks/25776884076/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test result: ok. 989 passed; 3 ignored;


FAILED TEST: tests/ui-toml/macro_metavars_in_unsafe/default/test.rs
command: CLIPPY_CONF_DIR="tests/ui-toml/macro_metavars_in_unsafe/default" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-c5e0235647f00232.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-2964c17de9728d6a.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-e7859c049d6fc5af.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-e1fa4d45c0e68db6.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-ae846d6788d6a9f1.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-e9aa61e1e4ee057e.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-3864f64671f03a2f.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-facaf493acd2f9ad.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-5689c7175177147d.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-d199b1d0494505d0.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-82c47a8e64b8c220.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-f7a5efbd7c1bc532.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-e9d338b744af34e7.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui-toml/macro_metavars_in_unsafe/default" "tests/ui-toml/macro_metavars_in_unsafe/default/test.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui-toml/macro_metavars_in_unsafe/default/test.stderr` to the actual output
--- tests/ui-toml/macro_metavars_in_unsafe/default/test.stderr
+++ <stderr output>
+++ <stderr output>
-error: the feature `lint_reasons` has been stable since 1.80.0-dev and no longer requires an attribute to enable
-  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:2:24
-   |
-LL | #![feature(decl_macro, lint_reasons)]
-   |
-   = note: `-D stable-features` implied by `-D warnings`
-   = help: to override `-D warnings` add `#[allow(stable_features)]`
-
-
 error: this macro expands metavariables in an unsafe block
   --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:19:9
... 181 lines skipped ...
    = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
-error: aborting due to 15 previous errors
+error: aborting due to 14 previous errors
 



full stderr:
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:19:9
   |
LL | /         unsafe {
LL | |
LL | |             dbg!($v);
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
   = help: to override `-D warnings` add `#[allow(clippy::macro_metavars_in_unsafe)]`

error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:30:9
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:30:9
   |
LL | /         unsafe {
LL | |
LL | |             $r#mod;
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:42:13
   |
LL | /             unsafe {
LL | /             unsafe {
LL | |
LL | |                 $v;
LL | |             }
   | |_____________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:67:17
   |
LL | /                 unsafe {
LL | /                 unsafe {
LL | |
LL | |                     $v;
LL | |                 }
   | |_________________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:95:21
   |
LL |                     unsafe { $inside_unsafe }
LL |                     unsafe { $inside_unsafe }
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:110:17
   |
LL | /                 unsafe {
LL | /                 unsafe {
LL | |
LL | |                     $inside_unsafe;
LL | |                 }
   | |_________________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:122:9
   |
LL | /         unsafe {
LL | /         unsafe {
LL | |
LL | |             $inside_unsafe;
LL | |             let inside_unsafe = 1;
LL | |             inside_unsafe;
LL | |         }
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:137:9
   |
LL | /         unsafe {
LL | /         unsafe {
LL | |
LL | |             $inside_unsafe;
LL | |         }
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:146:13
   |
LL | /             unsafe {
LL | /             unsafe {
LL | |
LL | |                 $x;
LL | |             }
   | |_____________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:171:5
   |
LL | /     unsafe {
LL | /     unsafe {
LL | |
LL | |         $v;
LL | |     }
   | |_____^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:158:9
   |
LL | /         unsafe {
LL | /         unsafe {
LL | |
LL | |             $x;
LL | |         }
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:162:9
   |
LL | /         unsafe {
LL | /         unsafe {
LL | |
LL | |             $x;
LL | |             $y;
LL | |         }
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:222:9
   |
LL | /         unsafe {
LL | /         unsafe {
LL | |
LL | |             $v;
LL | |         }
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: this macro expands metavariables in an unsafe block
##[error]  --> tests/ui-toml/macro_metavars_in_unsafe/default/test.rs:232:9
   |
LL | /         unsafe {
LL | /         unsafe {
LL | |
LL | |             nested_macro_helper!($v);
LL | |             $v;
   | |_________^
   |
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = note: this allows the user of the macro to write unsafe code outside of an unsafe block
   = help: consider expanding any metavariables outside of this block, e.g. by storing them in a variable
   = help: ... or also expand referenced metavariables in a safe context to require an unsafe block at callsite
error: aborting due to 14 previous errors


full stdout:
```

</details>

xFrednet commented:
Hey @wesleywiser, this PR is now ready for review. I've only rebased, so the changes should still be identical.

The PR still has a placeholder in Clippy's code for the stabilization version. I would like to update this shortly before the merge.

bors commented:
:umbrella: The latest upstream changes (presumably #126865) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented:
Removing the lang nomination, as the stabilization was approved a while ago.

Nominating for compiler, as this is a large PR that accumulates merge conflicts. Hopefully the next time the merge conflicts are resolved, it'll be able to be merged.

Urgau commented on compiler/rustc_lint_defs/src/builtin.rs:
```diff
@@ -606,13 +606,13 @@ declare_lint! {
 declare_lint! {
-    /// The `unfulfilled_lint_expectations` lint detects lint trigger expectations
-    /// that have not been fulfilled.
+    /// The `unfulfilled_lint_expectations` lint warns if a lint expectation is
+    /// unfulfilled.
```
nit: technically "warns" instead of "detects" is sightly worse since the first one doesn't take into user defined levels (`#![allow(unfulfilled_lint_expectations)]` would not longer make the lint warn)

xFrednet replied:
I've changed this to:

> The `unfulfilled_lint_expectations` lint detects when a lint expectation is unfulfilled.

Urgau commented on src/doc/rustc/src/lints/levels.md:
```diff
@@ -33,6 +34,40 @@ But this code violates the `missing_docs` lint.
+    #[expect(unused_variables)]
+    let unused = "Everyone ignores me";
+
+    #[expect(unused_variables)]
+    let used = "I'm useful";
```
nit: maybe add a comment in the code as to make it more obvious which `#[expect]` is unfulfilled
```suggestion
    #[expect(unused_variables)] // unused_variables lint is not emitted
    let used = "I'm useful";    // the expectation is therefore unfulfilled
```

Urgau commented on src/doc/rustc/src/lints/levels.md:
````diff
@@ -33,6 +34,40 @@ But this code violates the `missing_docs` lint.
+  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
+```
+
+This level can only be defined via the `#[expect]` attribute and not via the
+console. Lints with the special 'force-warn' lint will still be emitted as usual.
````
```suggestion
This level can only be defined via the `#[expect]` attribute, there is no equivalent
flag. Lints with the special 'force-warn' level will still be emitted as usual.
```

Urgau commented on tests/ui/lint/lint-removed-cmdline.stderr:
```diff
@@ -26,5 +26,10 @@ LL | #[deny(warnings)]
    = note: `#[deny(unused_variables)]` implied by `#[deny(warnings)]`
 
-error: aborting due to 1 previous error; 3 warnings emitted
+warning: lint `raw_pointer_derive` has been removed: using derive with raw pointers is ok
+   |
+   = note: requested on the command line with `-D raw_pointer_derive`
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
```
(not blocking) any idea why these diagnostics (here and in other tests) are duplicated multiple times?

xFrednet replied:
No sadly not :thinking:

Urgau reviewed (approved):
Left some comments but otherwise the Compiler changes looks fine to me!

r=me on the compiler part

xFrednet commented:
@rust-lang/clippy The compiler part has been approved. Could someone check the Clippy changes, since they include a small refactoring?

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9665729571/job/26663568428) [(plain)](https://github.com/rust-lang/rust/commit/60e19b9ee44ab3a0ff14b6842ce41ce37a8e6086/checks/26663568428/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.439   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.457 Collecting boolean-py==4.0
#13 0.464   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.481 Collecting chardet==5.1.0
---
#13 3.657 Building wheels for collected packages: reuse
#13 3.658   Building wheel for reuse (pyproject.toml): started
#13 3.986   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.987   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.988   Stored in directory: /tmp/pip-ephem-wheel-cache-l_r71xjc/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.990 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.013   Attempting uninstall: setuptools
#13 4.013     Found existing installation: setuptools 59.6.0
#13 4.015     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.015     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.015     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.685 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.685 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.219 Collecting virtualenv
#13 5.272   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.488      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 26.7 MB/s eta 0:00:00
#13 5.542 Collecting platformdirs<5,>=3.9.1
#13 5.550   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.588 Collecting filelock<4,>=3.12.2
#13 5.595   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.615 Collecting distlib<1,>=0.3.7
#13 5.623   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.633      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 56.7 MB/s eta 0:00:00
#13 5.723 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.897 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      198592 kB
DirectMap2M:     8189952 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/d929a42a664c026167800801b26d734db925314f/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-d929a42a664c026167800801b26d734db925314f-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5149 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_builtin_macros/src/lib.rs:15: line not in alphabetical order
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
some tidy checks failed
Build completed unsuccessfully in 0:01:49
  local time: Tue Jun 25 16:01:55 UTC 2024
  network time: Tue, 25 Jun 2024 16:01:55 GMT
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9665877416/job/26664092571) [(plain)](https://github.com/rust-lang/rust/commit/b124b3666e2109d01c8ec02ecae6d2e965e4bbe8/checks/26664092571/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [ui] tests/ui/lint/expect-future_breakage-crash-issue-126521.rs stdout ----

error: Error: expected failure status (Some(1)) but received status Some(101).
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/expect-future_breakage-crash-issue-126521.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/expect-future_breakage-crash-issue-126521" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/expect-future_breakage-crash-issue-126521/auxiliary"
--- stderr -------------------------------
thread 'rustc' panicked at compiler/rustc_errors/src/lib.rs:1912:34:
internal error: entered unreachable code
stack backtrace:
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: <rustc_errors::Level>::to_str
   4: <rustc_errors::json::Diagnostic>::from_errors_diagnostic
   5: <alloc::vec::into_iter::IntoIter<rustc_errors::diagnostic::DiagInner> as core::iter::traits::iterator::Iterator>::try_fold::<alloc::vec::in_place_drop::InPlaceDrop<rustc_errors::json::FutureBreakageItem>, core::iter::adapters::map::map_try_fold<rustc_errors::diagnostic::DiagInner, rustc_errors::json::FutureBreakageItem, alloc::vec::in_place_drop::InPlaceDrop<rustc_errors::json::FutureBreakageItem>, core::result::Result<alloc::vec::in_place_drop::InPlaceDrop<rustc_errors::json::FutureBreakageItem>, !>, <rustc_errors::json::JsonEmitter as rustc_errors::emitter::Emitter>::emit_future_breakage_report::{closure#0}, alloc::vec::in_place_collect::write_in_place_with_drop<rustc_errors::json::FutureBreakageItem>::{closure#0}>::{closure#0}, core::result::Result<alloc::vec::in_place_drop::InPlaceDrop<rustc_errors::json::FutureBreakageItem>, !>>
   6: alloc::vec::in_place_collect::from_iter_in_place::<core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<rustc_errors::diagnostic::DiagInner>, <rustc_errors::json::JsonEmitter as rustc_errors::emitter::Emitter>::emit_future_breakage_report::{closure#0}>, rustc_errors::json::FutureBreakageItem>
   7: <rustc_errors::json::JsonEmitter as rustc_errors::emitter::Emitter>::emit_future_breakage_report
   8: <rustc_errors::DiagCtxtHandle>::emit_future_breakage_report
   9: <rustc_session::session::Session>::finish_diagnostics
  10: rustc_span::create_session_globals_then::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.81.0-nightly (a52242819 2024-06-25) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
end of query stack
------------------------------------------

```

</details>

xFrednet committed:
Prevent ICE from expected future breakage

xFrednet commented:
Hey @Urgau, the rebase revealed an ICE, which I fixed in the last commit. Could you review that one? I've also addressed your comments, in the previous one. Thank you for the previous :heart:

Urgau commented on compiler/rustc_errors/src/json.rs:
```diff
@@ -135,7 +135,12 @@ impl Emitter for JsonEmitter {
             .into_iter()
             .map(|mut diag| {
-                if diag.level == crate::Level::Allow {
+                // The `FutureBreakageItem` is collected and serialized.
+                // However, the `allow` and `expect` lint levels can't usually
+                // be serialized. The lint level is overwritten to allow the
+                // serialization again and force a lint emission.
+                // (This is an educated guess. I didn't originally add this)
+                if matches!(diag.level, crate::Level::Allow | crate::Level::Expect(..)) {
```
I think this is mostly right. I looked at the introduction of [those lines](https://github.com/rust-lang/rust/commit/23018a55d9735afbefb19fcec91db4b53fe917d8#diff-4709ec799c1d3ac3ab848fc853bcf0250fae27935fe67a394a4832da9d7fa9b2R139-R141) and they come straight from the introduction of future incompatibility system.

Interestingly the PR also introduced the `allow` level, which I think is because before that PR an allowed lint could never trigger a lint, since well it's allowed, but future incompatibility forces the lint emission even through it's current level is `allow`, and the same happens with `expect`.

And because the emitter system doesn't expect `allow` or `expect` as they don't normally make sense, we change (just before emiting the lint) it's level to `warn` as to make the emitter happy but also to avoid confused users.

I would therefore suggest changing the text to something like this:
```suggestion
                // Allowed or expected lints don't normally (by definition) emit a lint
                // but future incompat lints are special and are emitted anyway.
                //
                // So to avoid ICEs and confused users we "upgrade" the lint level for
                // those `FutureBreakageItem` to warn.
                if matches!(diag.level, crate::Level::Allow | crate::Level::Expect(..)) {
```

xFrednet replied:
This is amazing. Thank you for tracking down the actual source and suggesting this. :heart:

iamahuman replied:
This suggestion hasn't been applied up until r+. @Urgau are you okay with this? Is this suggestion optional?

Urgau replied:
Um... no, I expect this suggestion to be applied.
Let me do it from the GitHub UI.

blyxyas reviewed (approved):
Everything looks alright to me from the Clippy side.

xFrednet commented:
Nice, the rustdoc and other parts have been approved earlier AFAIK. So that means we should be able to finally r+ this :tada: Thank you, to everyone who helped with this feature :tada: 

I'm r+ing this as of [this comment](https://github.com/rust-lang/rust/pull/120924#pullrequestreview-2136650960)

@bors r=Urgau,blyxyas

bors commented:
:pushpin: Commit 1d667a0937f6242366d1f47d359a2c739f3c62b1 has been approved by `Urgau,blyxyas`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Urgau,blyxyas 1d667a0937f6242366d1f47d359a2c739f3c62b1 -->
<!-- homu: {"type":"Approved","sha":"1d667a0937f6242366d1f47d359a2c739f3c62b1","approver":"Urgau,blyxyas","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Urgau committed:
Clarify comment on changing to warn future breakage items

https://github.com/rust-lang/rust/pull/120924/files#r1653512240

Urgau commented:
@bors r=Urgau,blyxyas

bors commented:
:pushpin: Commit 7666534381042f5febdc995e024af051ceecc2f5 has been approved by `Urgau,blyxyas`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Urgau,blyxyas 7666534381042f5febdc995e024af051ceecc2f5 -->
<!-- homu: {"type":"Approved","sha":"7666534381042f5febdc995e024af051ceecc2f5","approver":"Urgau,blyxyas","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

apiraino commented:
removing the compiler nomination as not necessary anymore

@rustbot label -I-compiler-nominated

bors commented:
:hourglass: Testing commit 7666534381042f5febdc995e024af051ceecc2f5 with merge 4bc39f028d14c24b04dd17dc425432c6ec354536...
<!-- homu: {"type":"BuildStarted","head_sha":"7666534381042f5febdc995e024af051ceecc2f5","merge_sha":"4bc39f028d14c24b04dd17dc425432c6ec354536"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9683477829/job/26724852083)
Approved by: Urgau,blyxyas
Pushing 4bc39f028d14c24b04dd17dc425432c6ec354536 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Urgau,blyxyas","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9683477829/job/26724852083"},"merge_sha":"4bc39f028d14c24b04dd17dc425432c6ec354536"} -->

rust-timer commented:
Finished benchmarking commit (4bc39f028d14c24b04dd17dc425432c6ec354536): [comparison URL](https://perf.rust-lang.org/compare.html?start=d7c59370cea68cd17006ec3440a43254fd0eda7d&end=4bc39f028d14c24b04dd17dc425432c6ec354536&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d7c59370cea68cd17006ec3440a43254fd0eda7d&end=4bc39f028d14c24b04dd17dc425432c6ec354536&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6% | [0.2%, 1.9%] | 142   |
| Regressions ❌ <br /> (secondary)  | 0.6% | [0.1%, 1.5%] | 79    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.6% | [0.2%, 1.9%] | 142   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d7c59370cea68cd17006ec3440a43254fd0eda7d&end=4bc39f028d14c24b04dd17dc425432c6ec354536&stat=max-rss)
<details>
<summary>Results (primary -3.8%, secondary -3.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.8% | [-3.8%, -3.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.8% | [-3.8%, -3.8%] | 1     |
| All ❌✅ (primary)                 | -3.8% | [-3.8%, -3.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d7c59370cea68cd17006ec3440a43254fd0eda7d&end=4bc39f028d14c24b04dd17dc425432c6ec354536&stat=cycles:u)
<details>
<summary>Results (primary 1.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.4% | [0.8%, 2.3%] | 21    |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.4% | [0.8%, 2.3%] | 21    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d7c59370cea68cd17006ec3440a43254fd0eda7d&end=4bc39f028d14c24b04dd17dc425432c6ec354536&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 695.122s -> 696.035s (0.13%)
**Artifact size**: 326.61 MiB -> 326.75 MiB (0.04%)

jhpratt commented:
How could this have caused so many regressions? It's not as if everyone is using the attribute at this point in time.

xFrednet commented:
It might be that the expectations were previously only collected/determined if the feature was enabled. That would be my guess, at least.

Nilstrieb commented:
it would be good to figure out the cause of the regression and whether it can be mitigated. if it's inherent to the feature, then we'll have to accept it

qarmin commented on src/tools/clippy/clippy_config/src/msrvs.rs:
```diff
@@ -17,6 +17,7 @@ macro_rules! msrv_aliases {
 
 // names may refer to stabilized feature flags or library items
 msrv_aliases! {
+    1,81,0  { LINT_REASONS_STABILIZATION }
```
One space too much after 0

Nilstrieb commented:
i checked the `exa` regression with cachegrind:
```
   3,586,341    <rustc_lint::levels::LintLevelsBuilder<rustc_lint::levels::QueryMapExpectationsWrapper>>::add_id
   1,119,278    <rustc_middle::lint::ShallowLintLevelMap>::lint_level_id_at_node
     676,462    rustc_hir::intravisit::walk_expr::<rustc_lint::levels::LintLevelsBuilder<rustc_lint::levels::QueryMapExpectationsWrapper>>

```

these are the functions that executed the most additional instructions and are therefore likely the cause of the regression.

pnkfelix commented:
Visiting for weekly rustc perf triage.

* wide collection of regressions.
* PR discussion indicates regression may be inherent to how `#[expect]` is implemented; it is also hypothesized to be "likely" that the implementation can be better optimized.
* not marking as triaged.

(my hope is that someone will look into the "further optimizations" that xFrednet [alludes to above](https://github.com/rust-lang/rust/pull/120924#issuecomment-2202486203), and after we've done a reasonable amount of investigation there, then we can mark this as triaged.)

pnkfelix commented:
Update browser-ui-test version to `0.18.0` [#127010](https://github.com/rust-lang/rust/pull/127010) [(Comparison Link)](https://perf.rust-lang.org/compare.html?start=9c3bc805dd9cb84019c124b9a50fdff1e62a7ec9&end=42add88d2275b95c98e512ab680436ede691e853&stat=instructions:u)

| (instructions:u)                   | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 7.2% | [7.2%, 7.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

myrrlyn commented:
As the (embarrassingly absent) RFC author, I am _thrilled_ to see this land! Thank you all for your effort in actually taking my initial ideas and doing all the effort to refine, implement, and ship them. I can't wait to start using it.
